File: recursion/leetcode_Spiral_Matrix.java
/*Spiral_Matrix

Given a matrix of m x n elements (m rows, n columns), return all elements of
the matrix in spiral order.
For example,
Given the following matrix:
[
  [ 1, 2, 3 ],
  [ 4, 5, 6 ],
  [ 7, 8, 9 ]
]
You should return [1,2,3,6,9,8,7,4,5].
*/

import java.util.ArrayList;


public class leetcode_Spiral_Matrix {

    public static void main(String[] args) {
		// Test case.
        int[][] test = new int[2][1];
        test[0][0] = 3;
        test[1][0] = 2;
        spiralOrder(test);
    }

    public static ArrayList<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer> res = new ArrayList<Integer>();
        if (matrix.length == 0) {
            return res;
        }
        int[][] onion = matrix;
        peelMatrix(onion, res);
        return res;
    }

    public static void peelMatrix(int[][] onion, ArrayList<Integer> res) {
        int row = onion.length;
        int col = onion[0].length;
        for (int i = 0; i < col; ++i) {
            res.add(onion[0][i]); // Peel the first row.
        }
        if (row > 1) {
            int[][] newOnion = new int[col][row - 1];
            for (int i = 0; i < row - 1; ++i) {
                for (int j = 0; j < col; ++j) {
                    // Rotate the the rest of matrix by 90 degree anti-clockwise.
                    newOnion[col - 1 - j][i] = onion[i + 1][j];
                }
            }
            peelMatrix(newOnion, res);
        }
    }
}

/* Python Version

def spiralOrder(self, matrix):
    if not matrix:
        return []

    n = len(matrix)
    m = len(matrix[0])

    ret = []
    x = y = 0

    while m >= 1 and n >= 1:
        if m == 1:
            # Tall.
            for i in xrange(x, x + n):
                ret.append(matrix[i][y + m - 1])
        elif n == 1:
            # flattened
            ret += matrix[x][y : y + m]
        else:
			# Common cases.

            # up
            ret += matrix[x][y : y + m]

            # right
            for i in xrange(x + 1, x + n - 1):
                ret.append(matrix[i][y + m - 1])

            # down
            ret += matrix[y + n - 1][y : y + m][::-1]

            # left
            for i in reversed(xrange(x + 1, x + n - 1)):
                ret.append(matrix[i][y])

        x += 1
        y += 1
        n -= 2
        m -= 2

    return ret
*/

File: recursion/leetcode_Spiral_Matrix.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode spiral matrix
 */

#include <stdio.h>
#include <vector>

using std::vector;


class Solution {
 public:
  /**
   * To avoid overpeeling, conditionally peel lower and left.
   */
  vector<int> spiralOrder(const vector<vector<int>>& matrix) {
    vector<int> res;
    if (matrix.size() == 0) {
      return res;
    }
    peel(&res, matrix, 0, 0, matrix.size() - 1, matrix[0].size() - 1);
    return res;
  }

 private:
  void peel(vector<int>* res, const vector<vector<int>>& matrix, int i , int j, int h, int k) {
    if (i <= h && j <= k) {
      // Peel upper.
      for (int ii = j; ii <= k; ++ii) {
        res->push_back(matrix[i][ii]);
      }

      // Peel right.
      for (int ii = i + 1; ii <= h; ++ii) {
        res->push_back(matrix[ii][k]);
      }

      if (i < h && j < k) {
        // Peel lower.
        for (int ii = k - 1; ii >= j; --ii) {
          res->push_back(matrix[h][ii]);
        }
        // Peel left.
        for (int ii = h - 1; ii > i; --ii) {
          res->push_back(matrix[ii][j]);
        }
        peel(res, matrix, i + 1, j + 1, h - 1, k - 1);
      }
    }
  }
};


int main() {
  Solution sol;
  vector<vector<int>> matrix {
    {1, 2, 3},
    {4, 5, 6},
    {7, 8, 9},
    {10, 11, 12},
  };
  vector<int> res = sol.spiralOrder(matrix);

  for (int val : res) {
    printf("%d ", val);
  }
  printf("\n");
}
